### PR TITLE
[Java] Add struts

### LINKED ISSUE #2976: [Java] Add struts
Hi @lukaszlenart @JCgH4164838Gh792C124B5

Would you mind to add (or consent that we add) [struts](https://struts.apache.org/) in `here` ?

This project compare performance based on 3 routes :

+ `GET` `/`, **SHOULD** return a successful status code and an empty string
+ `GET` `/user/:id`, **SHOULD** return a successful status code and the id
+ `POST` `/user`, **SHOULD** return a successful status code and an empty string

Regards,

### PR DESCRIPTION
**Pull request by lukaszlenart**
Closes #2976

### CODE REVIEW

File: java/struts2/config.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+framework:
+  website: struts.apache.org
+  version: 2.5.22
```
Comment (waghanza) [inline java/struts2/config.yaml:3]: could you use `X.Y` format (`2.5`), since it's clear for readers

File: java/struts2/pom.xml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>org.apache.struts.benchmarker</groupId>
+    <artifactId>struts2</artifactId>
+    <version>1.0-SNAPSHOT</version>
+    <packaging>war</packaging>
+    <name>struts2</name>
+
+    <properties>
+        <struts2.version>2.5.22</struts2.version>
```
Comment (waghanza) [inline java/struts2/pom.xml:12]: could you use an interval (I mean `[2.5,2.6)`)  ?

File: java/struts2/pom.xml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        <log4j2.version>2.13.3</log4j2.version>
```
Comment (waghanza) [inline java/struts2/pom.xml:13]: is this required ? I mean to specify a version

File: java/struts2/pom.xml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.8.1</version>
+                <configuration>
+                    <encoding>UTF-8</encoding>
+                    <source>1.8</source>
```
Comment (waghanza) [inline java/struts2/pom.xml:61]: :+1:

File: java/struts2/pom.xml [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                    <target>1.8</target>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.eclipse.jetty</groupId>
+                <artifactId>jetty-maven-plugin</artifactId>
+                <version>9.4.30.v20200611</version>
```
Comment (waghanza) [inline java/struts2/pom.xml:68]: is this required ? I mean to specify a version ?

**Comment by waghanza** [comment]
Thanks for you contribution :tada:  I have ask you some question to guarantee that all `java` **frameworks** run the same (at least the max possible) way.

PS : I've no `java` skill, so some questions could appears coming from a noob 

PSS : Could you also adds `structs2` to `.travis.yml` ? btw, why `struts2` and not `struts` ?

File: java/struts2/pom.xml (hunk 2 above)
Comment (lukaszlenart) [reply java/struts2/pom.xml:12]: I tried but then Maven uses `2.6-SNAPSHOT` version which is a development version. Being specific in Maven based project is a good approach.

File: java/struts2/pom.xml (hunk 3 above)
Comment (lukaszlenart) [reply java/struts2/pom.xml:13]: I can throw away the whole logging and just use what's provided.

File: java/struts2/config.yaml (hunk 1 above)
Comment (lukaszlenart) [reply java/struts2/config.yaml:3]: 👍

File: java/struts2/pom.xml (hunk 5 above)
Comment (lukaszlenart) [reply java/struts2/pom.xml:68]: Yes

**Comment by lukaszlenart** [comment]
Done, `struts2` is a different framework then `struts1` 

https://en.wikipedia.org/wiki/Apache_Struts_1
https://en.wikipedia.org/wiki/Apache_Struts_2

File: java/struts2/pom.xml (hunk 2 above)
Comment (waghanza) [reply java/struts2/pom.xml:12]: Ah ... I haven't the same problem with other `java` frameworks, but I have with `clojure` and `kotlin`. There is not way to specify a range **BUT** avoid any `development`, `rc` ... versions in **maven** ?

File: java/struts2/pom.xml (hunk 3 above)
Comment (waghanza) [reply java/struts2/pom.xml]: yeah, it will be awesome

The logging will be **disabled** for all frameworks

File: java/struts2/pom.xml (hunk 2 above)
Comment (lukaszlenart) [reply java/struts2/pom.xml:12]: It supposed to ignore `-SNAPSHOT` versions but somehow Maven includes it, I was testing locally and still the same.

```
Resolution of dependency ranges should not resolve to a snapshot (development version) unless it is included as an explicit boundary.
```

Comment (lukaszlenart) [reply java/struts2/pom.xml:12]: `mvn clean package`

```
dev-1:[1.8] ~/P/O/w/j/struts2 (master) > ll target/struts2-1.0/WEB-INF/lib
total 10936
-rw-r--r--+ 1 lukaszlenart  staff    72446 Apr 10 07:38 commons-fileupload-1.4.jar
-rw-r--r--+ 1 lukaszlenart  staff   214788 Apr 10 07:38 commons-io-2.6.jar
-rw-r--r--+ 1 lukaszlenart  staff   523372 Apr 19 10:23 commons-lang3-3.10.jar
-rw-r--r--+ 1 lukaszlenart  staff   207030 Apr 10 07:38 commons-text-1.8.jar
-rw-r--r--+ 1 lukaszlenart  staff  1702975 Apr 19 10:23 freemarker-2.3.30.jar
-rw-r--r--+ 1 lukaszlenart  staff   777727 Apr 10 07:39 javassist-3.24.1-GA.jar
-rw-r--r--+ 1 lukaszlenart  staff   292301 May 16 08:19 log4j-api-2.13.3.jar
-rw-r--r--+ 1 lukaszlenart  staff   264012 Apr 19 10:23 ognl-3.2.14.jar
-rw-r--r--+ 1 lukaszlenart  staff  1530698 Jun 22 08:12 struts2-core-2.6-SNAPSHOT.jar
```

Comment (lukaszlenart) [reply java/struts2/pom.xml]: Just added the version range as it works with 2.6 as well.

Comment (waghanza) [reply java/struts2/pom.xml]: I was tricked liked that before (either with `maven` or `gradle`)

File: java/struts2/config.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+framework:
+  website: struts.apache.org
+  version: 2.5
+
+build:
+  mvn clean package
```
Comment (waghanza) [inline java/struts2/config.yaml:6]: must be an array (missing a **-**)

**Comment by waghanza** [comment]
It will run but

+ we have `\n` for get /  
+ we have `\n\n0` for get /user/0
+ post on /user is failing

**Comment by lukaszlenart** [comment]
Should be fixed

**Comment by waghanza** [comment]
may I add you to `CODEOWNER` @lukaszlenart ? so as to ping you in case of modification of `struts2` here

File: java/struts2/src/main/webapp/WEB-INF/user-id.jsp [hunk 7]
```diff
@@ -0,0 +1,3 @@
+<%@ taglib prefix="s" uri="/struts-tags" %>
+<%@ page contentType="text/text;charset=UTF-8" %>
```
[the remark this answers is not in the document]
Comment (lukaszlenart) [reply java/struts2/src/main/webapp/WEB-INF/user-id.jsp:2]: Right, I overlooked that :(

**Comment by lukaszlenart** [comment]
> may I add you to `CODEOWNER` @lukaszlenart ? so as to ping you in case of modification of `struts2` here

Sure :)

**Comment by waghanza** [comment]
@lukaszlenart can we create a single binary or package embedding all dependencies ? (I think it's a **fatjar** or **uberjar** that I need)

The main idea is to avoid relying on a binary package for `jetty-runner`, but embed into the `pom.xml` (ensure the version ...) and it could :+1: for next version (on `cloud`)

**Comment by lukaszlenart** [comment]
You would like to avoid using `jetty-runner` and have everything embedded in one fat jar? And the run it with `java -jar target/struts2.jar`?

**Pull request merged**